STM32: Add STM32C092 SoC support and NUCLEO-C092RC

str4t0m opened the pull request:
Add STM32C091/STM32C092 as a member of the STM32C0 family with additional timer, serial interfaces and CAN-FD(C092 only) support.

- soc: st: stm32: add stm32c091/c092 support
- dt-bindings: clock: stm32c0: add can-fd domain clock helper
- dts: arm: st: stm32c0: add stm32c091/c092 support
- dts: arm: st: stm32c0:  add counter nodes to all timers
- Add overlays for the ADC API, UART async API, the counter basic API tests, as well as the st wkup-pin sample.
- boards: st: add nucleo_c092rc

str4t0m committed:
soc: st: stm32: add stm32c09x support

Adds SoC support for the STM32C091, and the STM32C092 SoCs
which are part of the STM32C0 series.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
dts: arm: st: stm32c0: add stm32c09x support

Add dts support for the STM32C091 and STM32C092 SoCs,
that are part of the STM32C0 series.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
boards: st: add nucleo_c092rc

Adds definitions for the nucleo_c092rc supporting
the STM32C092RC entry-level MCU with CAN-FD support.

The series does not have a pll, so with the given 48MHz
CAN core clock frequency the internal timing calculation
can't find parameters for 5MHz data bitrate,
and for 8MHz they have an error of 84/1000.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
dt-bindings: clock: stm32c0: add can-fd domain clock helper

On STM32C092 variants CAN-FD is supported, and the clock
source can be chosen in the CCIPR1 register.
This commit adds the helper to prepare that choice for the register.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
tests: drivers: adc: adc_api: add overlay for nucleo_c092rc

Add an overlay to use adc1 channel 0 (on PA0) for testing.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
tests: drivers: uart: uart_async_api: add nucleo_c092rc overlay

Add a overlay for the nucleo_c092 testing usart4 and dma
channels 6 and 7. These are all not available in smaller STM32C0 SoCs.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
dts: arm: st: stm32c0: add counter nodes to all timers

Add the counter nodes (compat st,stm32-counter)
to all timers of the STM32C0 series.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
tests: drivers: counter: counter_basic_api: add nucleo_c092rc overlay

Add a overlay to the counter_basic_api test to enable the counter node
of TIM2, TIM3, and TIM14 to TIM17.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

kartben commented on boards/st/nucleo_c092rc/doc/index.rst:
```diff
@@ -0,0 +1,126 @@
+
+.. zephyr:board-supported-runners::
+
+Nucleo C092RE board includes an ST-LINK/V2-1 embedded debug tool interface.
```
```suggestion
Nucleo C092RC board includes an ST-LINK/V2-1 embedded debug tool interface.
```

Please check for similar typos you might have elsewhere

str4t0m replied:
Thanks for spotting the typo. Should be fixed now.

mathieuchopstm reviewed (commented):
There is a typo in message for commit `tests: drivers: uart: uart_async_api: add nucleo_c092rc overlay` (`STM32CO` instead of `STM32C0`).

The following commits should come after the board is introduced:
* `tests: drivers: adc: adc_api: add overlay for nucleo_c092rc`
* `tests: drivers: uart: uart_async_api: add nucleo_c092rc overlay`
* `tests: drivers: counter: counter_basic_api: add nucleo_c092rc overlay`

\+ some remarks but otherwise LGTM on surface

str4t0m committed:
samples: boards: st: power_mgmt: wkup_pins: add nucleo_c092rc

Add an overlay to the st specific power_mgmt wkup_pins sample
for the nucleo_c092rc board.
The user_button on pc13 is used as the wakeup-src.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m commented on dts/arm/st/c0/stm32c092.dtsi:
```diff
@@ -0,0 +1,78 @@
+
+#include <st/c0/stm32c071.dtsi>
+
+/* STM32C092 is a superset of the C071 with the exception of USB support */
```
I somehow thought the C092 would have USB like the C071, and didn't want to add any definitions I couldn't test.
But given C092 just adds FDCAN I will add C091 as well.

mathieuchopstm commented on boards/st/nucleo_c092rc/nucleo_c092rc.dts:
Per table, ` * STM32C09xRx |   PC13  |   PC5   |   PB5   |` so all three pins can be added (even if only one - or none - is used).

mathieuchopstm commented on boards/st/nucleo_c092rc/nucleo_c092rc.dts:
```diff
@@ -0,0 +1,199 @@
+
+&pwr {
+	wkup-pin@2 {
+		reg = <0x2>;
+			wkup-gpios = <&gpioc 13 STM32_PWR_WKUP_PIN_NOT_MUXED>;
```
```suggestion
		reg = <0x2>;
		wkup-gpios = <&gpioc 13 STM32_PWR_WKUP_PIN_NOT_MUXED>;
```

str4t0m replied:
Oh, sorry for that oversight. Should be fixed now.

mathieuchopstm commented on boards/st/nucleo_c092rc/nucleo_c092rc.dts:
```diff
@@ -0,0 +1,199 @@
+			wkup-gpios = <&gpioc 13 STM32_PWR_WKUP_PIN_NOT_MUXED>;
+	};
+	wkup-pin@5 {
+		reg = <0x5>;
+			wkup-gpios = <&gpioc 5 STM32_PWR_WKUP_PIN_NOT_MUXED>;
```
```suggestion
		reg = <0x5>;
		wkup-gpios = <&gpioc 5 STM32_PWR_WKUP_PIN_NOT_MUXED>;
```

mathieuchopstm commented on boards/st/nucleo_c092rc/nucleo_c092rc.dts:
```diff
@@ -0,0 +1,199 @@
+			wkup-gpios = <&gpioc 5 STM32_PWR_WKUP_PIN_NOT_MUXED>;
+	};
+	wkup-pin@6 {
+		reg = <0x6>;
+			wkup-gpios = <&gpiob 5 STM32_PWR_WKUP_PIN_NOT_MUXED>;
```
```suggestion
		reg = <0x6>;
		wkup-gpios = <&gpiob 5 STM32_PWR_WKUP_PIN_NOT_MUXED>;
```

mathieuchopstm commented on dts/arm/st/c0/stm32c092.dtsi:
```diff
@@ -0,0 +1,24 @@
+			reg-names = "m_can", "message_ram";
+			interrupts = <30 0>, <31 0>;
+			interrupt-names = "int0", "int1";
+			clocks = <&rcc STM32_CLOCK(APB1, 12U)>;
```
U suffix should not be used in DTS (yes, the entire file uses it...... I really ought to fix this one day)
```suggestion
			clocks = <&rcc STM32_CLOCK(APB1, 12)>;
```

mathieuchopstm commented on dts/arm/st/c0/stm32c091.dtsi:
```diff
@@ -0,0 +1,67 @@
+		timers15: timers@40014000 {
+			compatible = "st,stm32-timers";
+			reg = <0x40014000 0x400>;
+			clocks = <&rcc STM32_CLOCK(APB1_2, 16U)>;
+			resets = <&rctl STM32_RESET(APB1H, 16U)>;
```
```suggestion
			clocks = <&rcc STM32_CLOCK(APB1_2, 16)>;
			resets = <&rctl STM32_RESET(APB1H, 16)>;
```

mathieuchopstm reviewed (commented):
Minor cosmetic remarks but LGTM otherwise.

mathieuchopstm reviewed (commented):
A few other cosmetic things I missed in previous review

erwango reviewed (changes requested):
Minor comment otherwise LGTM

str4t0m commented on dts/arm/st/c0/stm32c091.dtsi:
```diff
@@ -0,0 +1,67 @@
+#include <st/c0/stm32c071.dtsi>
+
+/* STM32C091 is a superset of the C071 with the exception of USB support */
+/delete-node/ &usb;
+/delete-node/ &usb_fs_phy;
```
I have added a comment to the stm32c091.dtsi file.
If you prefer to factor out the peripherals from the stm32c071.dtsi file, i.e. introduce the c051 line let me know.
I can't test it on stm32c051 specifically, but I don't think there would be much risk, given the errata of c051 doesn't list anything the c071 doesn't
